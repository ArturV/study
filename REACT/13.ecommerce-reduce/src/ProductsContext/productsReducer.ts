import type { TCartProduct } from "../types";
import type { TProductState } from "./types";

export const productsReducer = (
  state: TProductState,
  action: any
): TProductState => {
  switch (action.type) {
    case "setProducts": {
      return { ...state, products: action.payload };
    }

    case "addProductToCart": {
      const modifiedCartProducts = [...state.cartProducts];
      const cartProduct = modifiedCartProducts.find(
        (modifiedCartProduct) => modifiedCartProduct.id === action.payload.id
      );

      if (cartProduct) {
        cartProduct.amount++;
      } else {
        const newCartProduct = state.products.find(
          (item) => item.id === action.payload.id
        ) as TCartProduct;

        console.info({ newCartProduct });
        modifiedCartProducts.push({ ...newCartProduct, amount: 1 });
      }

      return {
        ...state,
        cartProducts: modifiedCartProducts,
      };
    }
    default:
      return state;
  }
};
