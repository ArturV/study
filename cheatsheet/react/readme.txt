//how to start project:

npx create-react-app react-app --template typescript .

// cd .. && npx create-react-app --template typescript // pakilam vienu katalogu auksciau ir irasom react


// to run:

npm start

// kodo galutine versija pasiekiama vartotojui:

npm run build

****
testing, types, typescript - is package.json perkeliame i devdependencies

"devDependencies": {
},


istrinam: reportWebwitals.js - istrinam ir is inde.tsx istrinam reportWebvitals
istrinam is package.json - webVitals is dependecies

kai pakeiciame package.json  !! SVARBU

rasome i konsole: npm i

dar geriau istrinti pries tai package-lock.json ir node_modules
****


npm i react-router-dom // route'us kurti biblioteka

npm i styled-components  // stylus kurti https://www.npmjs.com/package/styled-components
npm i --save-dev @types/styled-components

komentavimas: ctrl+ slash (ctrl+/)

//shortid to keys
npm i --save-dev @types/shortid

***************************


//testavimai:

npm i -D @testing-library/react

describe ("kuri komponenta testuojame", ()=>{

}) 

npm test - konsoleje

****************

coPilot - iskiepis 

sukuria elementa, tik tada kai paduotas title:

{props.title.length >= 1 && <h1>{props.title}</h1>}

///

<Button>Text1</Button>
<Button>Text2</Button>
<Button>Text3</Button>

tas pats:

["text1", "text2", "text3"].map(item => (
  <Button>{item}</Button>
))

********

FETCH:
https://blog.logrocket.com/modern-api-data-fetching-methods-react/

FETCH with AXIOS:
https://www.freecodecamp.org/news/fetch-data-react/

DELETE FROM ARRAY:
https://dev.to/collegewap/how-to-delete-an-item-from-the-state-array-in-react-kl


*******

storybook install:

npx sb init 
npm i -D @testing-library/dom

//paleisti:

npm run storybook


***************
Agile - metodologija kur galime viska valdyti

jira / azure devops

backlogs - sarasas darbu ka reikia padaryti

8 story point - 2sav
1 story point - 1d

confluence 
atlassian

3 amigos - svarbaus tiketo aptarimas

tesonet
hostinger
bankera
inulti
nordsecurity


action verbs