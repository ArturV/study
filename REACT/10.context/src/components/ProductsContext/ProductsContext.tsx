import { createContext } from "react";
import type { TProductsContext } from "./types";

const INITIAL_VALUE = {
  fetchedProducts: [],
  cartProducts: [],
  dispatch: () => {},
} as const; // neleidzia perrast reiksmiu (neveiks dispatch, push)

//INITIAL_VALUE.dispatch = () => {};
//INITIAL_VALUE.cartProducts.push(1);

export const ProductsContext = createContext<TProductsContext>(INITIAL_VALUE);
