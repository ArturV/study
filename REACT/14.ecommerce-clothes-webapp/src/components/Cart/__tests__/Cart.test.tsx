import { fireEvent, render, screen } from "@testing-library/react";
import { ProductsContext } from "../../ProductsContext";
import { Cart } from "../Cart";
import renderer from "react-test-renderer";

describe("Cart", () => {
  it("should render the cart with no items", () => {
    render(
      <ProductsContext.Provider
        value={{
          cartProducts: [],
          fetchedProducts: [],
          dispatch: () => {},
        }}
      >
        <Cart />
      </ProductsContext.Provider>
    );

    // expect(screen.getAllByLabelText("cart-product").length).toBe(0);

    expect(screen.getByRole("empty-cart-message")).toBeVisible();
    expect(screen.getByRole("empty-cart-message").textContent).toBe(
      "Your cart is empty"
    );
  });

  it("renders cart products and total price", () => {
    const cartProducts = [
      {
        id: 1,
        title: "Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops",
        price: 109.95,
        description:
          "Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday",
        category: "men's clothing",
        image: "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
        rating: { rate: 3.9, count: 120 },
        amount: 1,
      },
      {
        id: 2,
        title: "Mens Casual Premium Slim Fit T-Shirts ",
        price: 22.3,
        description:
          "Slim-fitting style, contrast raglan long sleeve, three-button henley placket, light weight & soft fabric for breathable and comfortable wearing. And Solid stitched shirts with round neck made for durability and a great fit for casual fashion wear and diehard baseball fans. The Henley style round neckline includes a three-button placket.",
        category: "men's clothing",
        image:
          "https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg",
        rating: { rate: 4.1, count: 259 },
        amount: 1,
      },
    ];

    render(
      <ProductsContext.Provider
        value={{
          cartProducts,
          fetchedProducts: [],
          dispatch: () => {},
        }}
      >
        <Cart />
      </ProductsContext.Provider>
    );

    expect(
      screen.getByText("Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops")
    ).toBeInTheDocument();
    expect(
      screen.getByText("Mens Casual Premium Slim Fit T-Shirts")
    ).toBeInTheDocument();
    expect(screen.getByText("Total price: €132.25")).toBeInTheDocument();
  });

  it("removes an item from the cart", () => {
    const cartProducts = [
      {
        id: 1,
        title: "Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops",
        price: 109.95,
        description:
          "Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday",
        category: "men's clothing",
        image: "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
        rating: { rate: 3.9, count: 120 },
        amount: 1,
      },
      {
        id: 2,
        title: "Mens Casual Premium Slim Fit T-Shirts ",
        price: 22.3,
        description:
          "Slim-fitting style, contrast raglan long sleeve, three-button henley placket, light weight & soft fabric for breathable and comfortable wearing. And Solid stitched shirts with round neck made for durability and a great fit for casual fashion wear and diehard baseball fans. The Henley style round neckline includes a three-button placket.",
        category: "men's clothing",
        image:
          "https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg",
        rating: { rate: 4.1, count: 259 },
        amount: 1,
      },
    ];

    const removeFunction = jest.fn();

    render(
      <ProductsContext.Provider
        value={{
          cartProducts,
          fetchedProducts: [],
          dispatch: removeFunction,
        }}
      >
        <Cart />
      </ProductsContext.Provider>
    );

    const removeButton = screen.getAllByLabelText("remove button")[0];
    fireEvent.click(removeButton);

    expect(removeFunction).toHaveBeenCalledWith({
      payload: { productId: 1 },
      type: "removeProduct",
    });

    expect(removeFunction).toHaveBeenCalledTimes(1);
  });

  it("displays the total price of all items in the cart", () => {
    const cartProducts = [
      {
        id: 1,
        title: "Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops",
        price: 109.95,
        description:
          "Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday",
        category: "men's clothing",
        image: "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
        rating: { rate: 3.9, count: 120 },
        amount: 1,
      },
      {
        id: 2,
        title: "Mens Casual Premium Slim Fit T-Shirts ",
        price: 22.3,
        description:
          "Slim-fitting style, contrast raglan long sleeve, three-button henley placket, light weight & soft fabric for breathable and comfortable wearing. And Solid stitched shirts with round neck made for durability and a great fit for casual fashion wear and diehard baseball fans. The Henley style round neckline includes a three-button placket.",
        category: "men's clothing",
        image:
          "https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg",
        rating: { rate: 4.1, count: 259 },
        amount: 1,
      },
    ];
    render(
      <ProductsContext.Provider
        value={{
          cartProducts,
          fetchedProducts: [],
          dispatch: () => {},
        }}
      >
        <Cart />
      </ProductsContext.Provider>
    );

    expect(screen.getByLabelText("total price")).toBeVisible();
    expect(screen.getByText("Total price: €132.25")).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const tree = renderer.create(<Cart />);

    expect(tree).toMatchSnapshot();
  });
});
