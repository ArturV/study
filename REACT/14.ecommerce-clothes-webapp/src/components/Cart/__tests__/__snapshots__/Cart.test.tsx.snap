// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart matches snapshot 1`] = `
<div
  className="MuiBox-root css-0"
>
  <ul />
  <div
    className="MuiBox-root css-1ez87es"
  >
    <h4
      className="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
    >
      Total price: 
    </h4>
    <h4
      className="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
    >
      €0.00
    </h4>
  </div>
</div>
`;
